src/viam/ur/module/ur_arm.cpp

```diff
@@ -897,34 +890,38 @@ URArm::UrDriverStatus URArm::read_joint_keep_alive_(bool log) {
     std::string status;
     try {
         if (current_state_->local_disconnect.load()) {
-            // sleep just so we don't spam
-            // this may be unnecessary
-            std::this_thread::sleep_for(k_estop_delay);
+            if (current_state_->dashboard->commandIsInRemoteControl()) {
+                // reconnect to the tablet. We have to do this, otherwise the client will assume that the arm is still in local mode.
+                // yes, even though the client can already recognize that we are in remote control mode
+                current_state_->dashboard->disconnect();
+                if (!current_state_->dashboard->connect()) {
+                    return UrDriverStatus::DASHBOARD_FAILURE;
+                }
+                // reset the driver client so we are not asking for data
+                current_state_->driver->resetRTDEClient(current_state_->appdir + k_output_recipe, current_state_->appdir + k_input_recipe);
+                // reset the primary client so the driver is aware the arm is back in remote mode
+                // this has to happen, otherwise when an estop is triggered the arm will return error code C210A0
+                current_state_->driver->stopPrimaryClientCommunication();
+                current_state_->driver->startPrimaryClientCommunication();
+
+                // check if we can still send trajectories. if not we need to send the robot program again.
+                if (!current_state_->driver->isReverseInterfaceConnected() || !current_state_->driver->isTrajectoryInterfaceConnected()) {
+                    // send control script again to complete the restart
+                    if (!current_state_->driver->sendRobotProgram()) {
+                        throw std::runtime_error(
+                            "read_joint_keep_alive driver->sendRobotProgram() returned false when attempting to reconnect to the arm");
+                    }
+                }
 
-            if (!current_state_->dashboard->commandIsInRemoteControl()) {
-                return UrDriverStatus::DASHBOARD_FAILURE;
-            }
+                // turn communication with the RTDE client back on so we can receive data from the arm
+                current_state_->driver->startRTDECommunication();
```

Comment:
don't like how this part of the code keeps growing with random reset commands. I don't think I can simplify this any further though